### PR TITLE
ATO-440: Account interventions reset password race condition

### PR DESCRIPTION
**Pull request by billrdunn**
## What

- Redefined account interventions class structure. Two immutable classes are defined which hold the the raw response from AIS (`AccountInterventionState` and `AccountInterventionDetails`). A further class, `AccountIntervention`, handles account interventions logic applicable to Orchestration.  The idea is that in future, a developer should only need to use `AccountIntervention` and none of the other classes in production code.
- Updated codebase to use new structure.
- Serialised more of the response from account intervention service into `AccountInterventionDetails` which includes `appliedAt` value (the time that the intervention was applied).
- Read `password_reset_time` claim from Authentication
- Implement fix for race condition: if `password_reset_time` is after `appliedAt` then ignore `passwordReset` on the intervention.

## How to review
1. Code Review - commit by commit will be easiest

## 🚨⚠️ Orchestration and Authentication mutual dependencies ⚠️ 🚨

- [x] Impact on orch and auth mutual dependencies has been checked.

## Related PRs
PR to send `password_reset_time` claim from Authentication: https://github.com/govuk-one-login/authentication-api/pull/4104

### CODE REVIEW

**Review by BeckaL — COMMENTED** [review]
Approving on the basis that this looks like the logic we agreed. Will let someone from your team have the final sign off :-)

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final AccountInterventionDetails details;
+    private final AccountInterventionState state;
+    private final AccountInterventionStatus status;
+    private final boolean ignoreResetPassword;
+
+    public AccountIntervention(AccountInterventionState state) {
+        this.details =
+                new AccountInterventionDetails(
+                        Long.MIN_VALUE,
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:13]: Would `0L` make more sense here? Feels weird to have negative time

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE,
+                        "",
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE);
+        this.state = state;
+        this.ignoreResetPassword = details.appliedAt() < passwordResetTime;
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:44]: This is always true, right?

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                new AccountInterventionDetails(
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE,
+                        "",
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE);
+        this.state = state;
+        this.ignoreResetPassword = false;
+        this.status = deduceStatus();
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:21]: Nitpick,but `calculate` feels more conventional than `deduce`

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java (hunk 1 above)
Comment (billrdunn) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:13]: I thought about that and then decided on `MIN_VALUE` as it's a bit safer.  Using `MIN_VALUE` means `ignorePasswordReset` will never be `true`

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE,
+                        "",
+                        Long.MIN_VALUE,
+                        Long.MIN_VALUE);
+        this.state = state;
+        this.ignoreResetPassword = details.appliedAt() < passwordResetTime;
+        this.status = deduceStatus();
+    }
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:46]: This constructor feels a little odd. Given it's only used in one test I think I'd construct the details object in the test and remove this constructor.

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java (hunk 2 above)
Comment (billrdunn) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:44]: Will remove this constructor

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java (hunk 3 above)
Comment (ethanmills) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/AccountIntervention.java:21]: Nah, fine to leave

**Review by ethanmills — APPROVED** [review]

**Pull request merged**